// backport/dist/lib/github/v4/validateTargetBranch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateTargetBranch = void 0;
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const BackportError_1 = require("../../BackportError");
const ora_1 = require("../../ora");
const apiRequestV4_1 = require("./apiRequestV4");
async function validateTargetBranch({ accessToken, repoName, repoOwner, branchName, githubApiBaseUrlV4 = 'https://api.github.com/graphql', interactive, }) {
    const query = (0, graphql_tag_1.default) `
    query GetBranchId(
      $repoOwner: String!
      $repoName: String!
      $branchName: String!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $branchName) {
          id
        }
      }
    }
  `;
    const spinner = (0, ora_1.ora)(interactive, '').start();
    const res = await (0, apiRequestV4_1.apiRequestV4)({
        githubApiBaseUrlV4,
        accessToken,
        query,
        variables: { repoOwner, repoName, branchName },
    });
    if (!res.repository.ref) {
        spinner.fail(`The branch "${branchName}" does not exist`);
        throw new BackportError_1.BackportError({
            code: 'invalid-branch-exception',
            branchName: branchName,
        });
    }
    spinner.stop();
    return;
}
exports.validateTargetBranch = validateTargetBranch;
