// backport/dist/lib/github/v3/addAssigneesToPullRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAssigneesToPullRequest = void 0;
const rest_1 = require("@octokit/rest");
const elastic_apm_node_1 = __importDefault(require("elastic-apm-node"));
const ora_1 = require("../../../lib/ora");
const logger_1 = require("../../logger");
async function addAssigneesToPullRequest({ 
// options
githubApiBaseUrlV3, repoName, repoOwner, accessToken, autoAssign, interactive, dryRun, 
// additional args
pullNumber, assignees, }) {
    const text = autoAssign
        ? `Self-assigning to #${pullNumber}`
        : `Adding assignees to #${pullNumber}: ${assignees.join(', ')}`;
    logger_1.logger.info(text);
    const spinner = (0, ora_1.ora)(interactive, text).start();
    if (dryRun) {
        spinner.succeed();
        return;
    }
    const span = elastic_apm_node_1.default.startSpan('REST: Add assignees');
    try {
        const octokit = new rest_1.Octokit({
            auth: accessToken,
            baseUrl: githubApiBaseUrlV3,
            log: logger_1.logger,
        });
        await octokit.issues.addAssignees({
            owner: repoOwner,
            repo: repoName,
            issue_number: pullNumber,
            assignees: assignees,
        });
        spinner.succeed();
    }
    catch (e) {
        spinner.fail();
        logger_1.logger.info(`Could not add assignees to PR ${pullNumber}`, e);
    }
    finally {
        span?.end();
    }
}
exports.addAssigneesToPullRequest = addAssigneesToPullRequest;
