// apidoc-markdown/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.templateUtils = exports.loadTemplate = exports.loadApidocFiles = exports.isInTemplatesDir = exports.loadFromCliParamOrApiDocProject = exports.loadFileOrThrowIfNotExist = exports.pathExists = exports.unique = exports.mkdirp = exports.DEFAULT_TEMPLATE_PATH = exports.TEMPLATES_PATH = void 0;
const ejs_1 = __importDefault(require("ejs"));
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
exports.TEMPLATES_PATH = path_1.default.resolve(__dirname, '..', 'templates');
exports.DEFAULT_TEMPLATE_PATH = path_1.default.resolve(exports.TEMPLATES_PATH, 'default.md');
/**
 * Recursively create an arborescence to the given path
 *
 * @param p Path to follow
 */
const mkdirp = (p) => fs_1.promises.mkdir(p, { recursive: true });
exports.mkdirp = mkdirp;
/**
 * Make array values unique
 *
 * @param arr Source array of strings
 */
const unique = (arr) => [...new Set(arr)];
exports.unique = unique;
/**
 * Check a path exists on the file system
 *
 * @param p Path to check existance from
 * @throws Path does not exist
 */
const pathExists = (p, logIfFail = true) => fs_1.promises
    .access(p)
    .then(() => true)
    .catch(err => (logIfFail && console.error(err), false));
exports.pathExists = pathExists;
const loadFileOrThrowIfNotExist = async (optionName, filePath) => {
    if (!(await exports.pathExists(filePath)))
        throw new Error(`The \`${optionName}\` path does not exist or is not readable. Path: ${filePath}`);
    return await fs_1.promises.readFile(filePath, { encoding: 'utf-8' });
};
exports.loadFileOrThrowIfNotExist = loadFileOrThrowIfNotExist;
/**
 * Try to load file from CLI parameter or from `api_project.json`
 * @param optionName CLI parameter/`api_project.json` key name
 * @param cliParam Received CLI param value
 * @param apiDocProjectData `api_project.json` content
 * @returns
 */
const loadFromCliParamOrApiDocProject = async (optionName, cliParam, apiDocProjectData) => {
    if (cliParam)
        return await exports.loadFileOrThrowIfNotExist(`cli.${optionName}`, cliParam);
    else if (apiDocProjectData[optionName]) {
        if (apiDocProjectData[optionName].filename)
            // This is not a valid option as apiDoc will convert input `filename` to actual markdown content
            // But let's support it anyway.
            return await exports.loadFileOrThrowIfNotExist(`apidoc_project_file.${optionName}.filename`, apiDocProjectData[optionName].filename);
        else if (apiDocProjectData[optionName].content)
            return `${apiDocProjectData[optionName].title ? `# ${apiDocProjectData[optionName].title}\n\n` : ''}${apiDocProjectData[optionName].content}`;
    }
};
exports.loadFromCliParamOrApiDocProject = loadFromCliParamOrApiDocProject;
/**
 * Check if a template is in the templates directory
 * @param name template name (without the `.md` file extension)
 */
const isInTemplatesDir = (name) => fs_1.promises.readdir(exports.TEMPLATES_PATH).then(files => files.includes(`${name}.md`));
exports.isInTemplatesDir = isInTemplatesDir;
/**
 * Load apiDoc files, be backward-compatible with legacy `apidoc.json`
 * @param apiDocPath Path to generated apiDoc output directory. Where `api_data.json` and `api_project.json` are located
 * @throws Could not load `api_project.json` or `apidoc.json`
 */
const loadApidocFiles = async (apiDocPath) => {
    const apiDocProjectData = await Promise.resolve().then(() => __importStar(require(path_1.default.resolve(apiDocPath, 'api_project.json')))).catch(() => Promise.resolve().then(() => __importStar(require(path_1.default.resolve(apiDocPath, 'apidoc.json')))))
        .catch(err => {
        err.message = 'Could not load `api_project.json` or `apidoc.json` - ' + err.message;
        throw err;
    });
    const apiDocApiData = Object.values(await Promise.resolve().then(() => __importStar(require(path_1.default.resolve(apiDocPath, 'api_data.json'))))).filter(x => x.type);
    return { apiDocProjectData, apiDocApiData };
};
exports.loadApidocFiles = loadApidocFiles;
/**
 * Load the template and compile it with EJS
 * @param template template path, name or raw plain text
 * @param logIfFileNotFound turn off logging
 */
const loadTemplate = async (template, logIfFileNotFound = true) => {
    if (template) {
        // Raw EJS plain text template
        if (template.includes('<%'))
            logIfFileNotFound = false;
        // Template name
        else if (await exports.isInTemplatesDir(template))
            template = await fs_1.promises.readFile(path_1.default.resolve(exports.TEMPLATES_PATH, `${template}.md`), 'utf-8');
        // Path to template file
        else if (await exports.pathExists(template, logIfFileNotFound))
            template = await fs_1.promises.readFile(template, 'utf-8');
    }
    // Default template
    else
        template = await fs_1.promises.readFile(exports.DEFAULT_TEMPLATE_PATH, 'utf-8');
    return ejs_1.default.compile(template, { async: true });
};
exports.loadTemplate = loadTemplate;
/** Utility functions passed to the EJS template */
exports.templateUtils = {
    /**
     * Convert a title to a Markdown-valid relative link
     * @param str Title to convert
     * @returns The valid linkable string
     */
    toLink: (str) => str.replace(/\s+/g, '-')
};
