// rxjs-marbles/tape/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fakeSchedulers = exports.marbles = exports.cases = exports.configure = void 0;
var tape = require("tape");
var cases_1 = require("../cases");
var configuration_1 = require("../configuration");
var fake_1 = require("../fake");
var marbles_1 = require("../marbles");
__exportStar(require("../configuration"), exports);
__exportStar(require("../context"), exports);
__exportStar(require("../expect"), exports);
__exportStar(require("../fake"), exports);
function configure(configuration) {
    var factory = function (t) { return ({
        assert: t.ok.bind(t),
        assertDeepEqual: t.deepEqual.bind(t),
    }); };
    var configured = marbles_1.configure(function (t) { return (__assign(__assign(__assign({}, configuration_1.defaults()), factory(t)), configuration)); });
    var marbles = configured.marbles;
    function cases(name, func, cases) {
        cases_1._cases(function (c) {
            var t = c.only ? tape.only : c.skip ? tape.skip : tape;
            t(name + " / " + c.name, marbles(function (m, t) { return func(m, c, t); }));
        }, cases);
    }
    return { cases: cases, marbles: marbles };
}
exports.configure = configure;
var _a = configure({}), cases = _a.cases, marbles = _a.marbles;
exports.cases = cases;
exports.marbles = marbles;
function fakeSchedulers(fakeTest) {
    return fake_1.fakeSchedulers(fakeTest);
}
exports.fakeSchedulers = fakeSchedulers;
