// @elastic/eui/es/components/button/button_display/_button_display_content.styles.js
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import { css } from '@emotion/react';
import { logicalCSS } from '../../../global_styling';

var _ref = process.env.NODE_ENV === "production" ? {
  name: "ske5nx-isDisabled",
  styles: "pointer-events:auto;cursor:not-allowed;label:isDisabled;"
} : {
  name: "ske5nx-isDisabled",
  styles: "pointer-events:auto;cursor:not-allowed;label:isDisabled;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var _ref2 = process.env.NODE_ENV === "production" ? {
  name: "gtg6j-euiButtonDisplayContent__icon",
  styles: "flex-shrink:0;label:euiButtonDisplayContent__icon;"
} : {
  name: "gtg6j-euiButtonDisplayContent__icon",
  styles: "flex-shrink:0;label:euiButtonDisplayContent__icon;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var _ref3 = process.env.NODE_ENV === "production" ? {
  name: "10wlo76-euiButtonDisplayContent__spinner",
  styles: "flex-shrink:0;label:euiButtonDisplayContent__spinner;"
} : {
  name: "10wlo76-euiButtonDisplayContent__spinner",
  styles: "flex-shrink:0;label:euiButtonDisplayContent__spinner;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var _ref4 = process.env.NODE_ENV === "production" ? {
  name: "qfl6yj-right",
  styles: "flex-direction:row-reverse;label:right;"
} : {
  name: "qfl6yj-right",
  styles: "flex-direction:row-reverse;label:right;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

export var euiButtonDisplayContentStyles = function euiButtonDisplayContentStyles(_ref5) {
  var euiTheme = _ref5.euiTheme;
  return {
    // Base
    euiButtonDisplayContent: /*#__PURE__*/css(logicalCSS('height', '100%'), ";", logicalCSS('width', '100%'), ";display:flex;justify-content:center;align-items:center;vertical-align:middle;gap:", euiTheme.size.s, ";;label:euiButtonDisplayContent;"),
    // Icon side
    left: /*#__PURE__*/css(";label:left;"),
    right: _ref4,
    euiButtonDisplayContent__spinner: _ref3,
    euiButtonDisplayContent__icon: _ref2,
    isDisabled: _ref
  };
};
