// @loaders.gl/shapefile/dist/esm/lib/parsers/parse-shapefile.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _awaitAsyncGenerator from "@babel/runtime/helpers/esm/awaitAsyncGenerator";
import _wrapAsyncGenerator from "@babel/runtime/helpers/esm/wrapAsyncGenerator";
import _asyncIterator from "@babel/runtime/helpers/esm/asyncIterator";

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Proj4Projection } from '@math.gl/proj4';
import { binaryToGeoJson, transformGeoJsonCoords } from '@loaders.gl/gis';
import { parseShx } from './parse-shx';
import { zipBatchIterators } from '../streaming/zip-batch-iterators';
import { SHPLoader } from '../../shp-loader';
import { DBFLoader } from '../../dbf-loader';
export function parseShapefileInBatches(_x, _x2, _x3) {
  return _parseShapefileInBatches.apply(this, arguments);
}

function _parseShapefileInBatches() {
  _parseShapefileInBatches = _wrapAsyncGenerator(_regeneratorRuntime.mark(function _callee(asyncIterator, options, context) {
    var _ref, _ref$reproject, reproject, _ref$_targetCrs, _targetCrs, parseInBatches, fetch, url, _yield$_awaitAsyncGen, shx, cpg, prj, shapeIterator, propertyIterator, dbfResponse, shapeHeader, dbfHeader, iterator, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _value, item, geometries, properties, _item, geojsonGeometries, features;

    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _ref = options && options.gis || {}, _ref$reproject = _ref.reproject, reproject = _ref$reproject === void 0 ? false : _ref$reproject, _ref$_targetCrs = _ref._targetCrs, _targetCrs = _ref$_targetCrs === void 0 ? 'WGS84' : _ref$_targetCrs;
            parseInBatches = context.parseInBatches, fetch = context.fetch, url = context.url;
            _context.next = 4;
            return _awaitAsyncGenerator(loadShapefileSidecarFiles(options, context));

          case 4:
            _yield$_awaitAsyncGen = _context.sent;
            shx = _yield$_awaitAsyncGen.shx;
            cpg = _yield$_awaitAsyncGen.cpg;
            prj = _yield$_awaitAsyncGen.prj;
            _context.next = 10;
            return _awaitAsyncGenerator(parseInBatches(asyncIterator, SHPLoader, options));

          case 10:
            shapeIterator = _context.sent;
            _context.next = 13;
            return _awaitAsyncGenerator(fetch(replaceExtension(url, 'dbf')));

          case 13:
            dbfResponse = _context.sent;

            if (!dbfResponse.ok) {
              _context.next = 18;
              break;
            }

            _context.next = 17;
            return _awaitAsyncGenerator(parseInBatches(dbfResponse, DBFLoader, _objectSpread(_objectSpread({}, options), {}, {
              dbf: {
                encoding: cpg || 'latin1'
              }
            })));

          case 17:
            propertyIterator = _context.sent;

          case 18:
            _context.next = 20;
            return _awaitAsyncGenerator(shapeIterator.next());

          case 20:
            shapeHeader = _context.sent.value;

            if (!(shapeHeader && shapeHeader.batchType === 'metadata')) {
              _context.next = 25;
              break;
            }

            _context.next = 24;
            return _awaitAsyncGenerator(shapeIterator.next());

          case 24:
            shapeHeader = _context.sent.value;

          case 25:
            dbfHeader = {};

            if (!propertyIterator) {
              _context.next = 34;
              break;
            }

            _context.next = 29;
            return _awaitAsyncGenerator(propertyIterator.next());

          case 29:
            dbfHeader = _context.sent.value;

            if (!(dbfHeader && dbfHeader.batchType === 'metadata')) {
              _context.next = 34;
              break;
            }

            _context.next = 33;
            return _awaitAsyncGenerator(propertyIterator.next());

          case 33:
            dbfHeader = _context.sent.value;

          case 34:
            if (!propertyIterator) {
              _context.next = 40;
              break;
            }

            _context.next = 37;
            return _awaitAsyncGenerator(zipBatchIterators(shapeIterator, propertyIterator));

          case 37:
            iterator = _context.sent;
            _context.next = 41;
            break;

          case 40:
            iterator = shapeIterator;

          case 41:
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _context.prev = 43;
            _iterator = _asyncIterator(iterator);

          case 45:
            _context.next = 47;
            return _awaitAsyncGenerator(_iterator.next());

          case 47:
            _step = _context.sent;
            _iteratorNormalCompletion = _step.done;
            _context.next = 51;
            return _awaitAsyncGenerator(_step.value);

          case 51:
            _value = _context.sent;

            if (_iteratorNormalCompletion) {
              _context.next = 65;
              break;
            }

            item = _value;
            geometries = void 0;
            properties = void 0;

            if (!propertyIterator) {
              geometries = item;
            } else {
              _item = _slicedToArray(item, 2);
              geometries = _item[0];
              properties = _item[1];
            }

            geojsonGeometries = parseGeometries(geometries);
            features = joinProperties(geojsonGeometries, properties);

            if (reproject) {
              features = reprojectFeatures(features, prj, _targetCrs);
            }

            _context.next = 62;
            return {
              encoding: cpg,
              prj: prj,
              shx: shx,
              header: shapeHeader,
              data: features
            };

          case 62:
            _iteratorNormalCompletion = true;
            _context.next = 45;
            break;

          case 65:
            _context.next = 71;
            break;

          case 67:
            _context.prev = 67;
            _context.t0 = _context["catch"](43);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 71:
            _context.prev = 71;
            _context.prev = 72;

            if (!(!_iteratorNormalCompletion && _iterator["return"] != null)) {
              _context.next = 76;
              break;
            }

            _context.next = 76;
            return _awaitAsyncGenerator(_iterator["return"]());

          case 76:
            _context.prev = 76;

            if (!_didIteratorError) {
              _context.next = 79;
              break;
            }

            throw _iteratorError;

          case 79:
            return _context.finish(76);

          case 80:
            return _context.finish(71);

          case 81:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[43, 67, 71, 81], [72,, 76, 80]]);
  }));
  return _parseShapefileInBatches.apply(this, arguments);
}

export function parseShapefile(_x4, _x5, _x6) {
  return _parseShapefile.apply(this, arguments);
}

function _parseShapefile() {
  _parseShapefile = _asyncToGenerator(_regeneratorRuntime.mark(function _callee2(arrayBuffer, options, context) {
    var _ref2, _ref2$reproject, reproject, _ref2$_targetCrs, _targetCrs, parse, _yield$loadShapefileS, shx, cpg, prj, _yield$parse, header, geometries, geojsonGeometries, properties, url, fetch, dbfResponse, features;

    return _regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _ref2 = options && options.gis || {}, _ref2$reproject = _ref2.reproject, reproject = _ref2$reproject === void 0 ? false : _ref2$reproject, _ref2$_targetCrs = _ref2._targetCrs, _targetCrs = _ref2$_targetCrs === void 0 ? 'WGS84' : _ref2$_targetCrs;
            parse = context.parse;
            _context2.next = 4;
            return loadShapefileSidecarFiles(options, context);

          case 4:
            _yield$loadShapefileS = _context2.sent;
            shx = _yield$loadShapefileS.shx;
            cpg = _yield$loadShapefileS.cpg;
            prj = _yield$loadShapefileS.prj;
            _context2.next = 10;
            return parse(arrayBuffer, SHPLoader, options);

          case 10:
            _yield$parse = _context2.sent;
            header = _yield$parse.header;
            geometries = _yield$parse.geometries;
            geojsonGeometries = parseGeometries(geometries);
            properties = [];
            url = context.url, fetch = context.fetch;
            _context2.next = 18;
            return fetch(replaceExtension(url, 'dbf'));

          case 18:
            dbfResponse = _context2.sent;

            if (!dbfResponse.ok) {
              _context2.next = 23;
              break;
            }

            _context2.next = 22;
            return parse(dbfResponse, DBFLoader, {
              dbf: {
                encoding: cpg || 'latin1'
              }
            });

          case 22:
            properties = _context2.sent;

          case 23:
            features = joinProperties(geojsonGeometries, properties);

            if (reproject) {
              features = reprojectFeatures(features, prj, _targetCrs);
            }

            return _context2.abrupt("return", {
              encoding: cpg,
              prj: prj,
              shx: shx,
              header: header,
              data: features
            });

          case 26:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _parseShapefile.apply(this, arguments);
}

function parseGeometries(geometries) {
  var geojsonGeometries = [];

  var _iterator2 = _createForOfIteratorHelper(geometries),
      _step2;

  try {
    for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
      var geom = _step2.value;
      geojsonGeometries.push(binaryToGeoJson(geom, geom.type, 'geometry'));
    }
  } catch (err) {
    _iterator2.e(err);
  } finally {
    _iterator2.f();
  }

  return geojsonGeometries;
}

function joinProperties(geometries, properties) {
  var features = [];

  for (var i = 0; i < geometries.length; i++) {
    var geometry = geometries[i];
    var feature = {
      type: 'Feature',
      geometry: geometry,
      properties: properties && properties[i] || {}
    };
    features.push(feature);
  }

  return features;
}

function reprojectFeatures(features, sourceCrs, targetCrs) {
  var projection = new Proj4Projection({
    from: sourceCrs || 'WGS84',
    to: targetCrs || 'WGS84'
  });
  return transformGeoJsonCoords(features, function (coord) {
    return projection.project(coord);
  });
}

export function loadShapefileSidecarFiles(_x7, _x8) {
  return _loadShapefileSidecarFiles.apply(this, arguments);
}

function _loadShapefileSidecarFiles() {
  _loadShapefileSidecarFiles = _asyncToGenerator(_regeneratorRuntime.mark(function _callee3(options, context) {
    var url, fetch, shxPromise, cpgPromise, prjPromise, shx, cpg, prj, shxResponse, arrayBuffer, cpgResponse, prjResponse;
    return _regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            url = context.url, fetch = context.fetch;
            shxPromise = fetch(replaceExtension(url, 'shx'));
            cpgPromise = fetch(replaceExtension(url, 'cpg'));
            prjPromise = fetch(replaceExtension(url, 'prj'));
            _context3.next = 6;
            return Promise.all([shxPromise, cpgPromise, prjPromise]);

          case 6:
            shx = null;
            cpg = null;
            prj = null;
            _context3.next = 11;
            return shxPromise;

          case 11:
            shxResponse = _context3.sent;

            if (!shxResponse.ok) {
              _context3.next = 17;
              break;
            }

            _context3.next = 15;
            return shxResponse.arrayBuffer();

          case 15:
            arrayBuffer = _context3.sent;
            shx = parseShx(arrayBuffer);

          case 17:
            _context3.next = 19;
            return cpgPromise;

          case 19:
            cpgResponse = _context3.sent;

            if (!cpgResponse.ok) {
              _context3.next = 24;
              break;
            }

            _context3.next = 23;
            return cpgResponse.text();

          case 23:
            cpg = _context3.sent;

          case 24:
            _context3.next = 26;
            return prjPromise;

          case 26:
            prjResponse = _context3.sent;

            if (!prjResponse.ok) {
              _context3.next = 31;
              break;
            }

            _context3.next = 30;
            return prjResponse.text();

          case 30:
            prj = _context3.sent;

          case 31:
            return _context3.abrupt("return", {
              shx: shx,
              cpg: cpg,
              prj: prj
            });

          case 32:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _loadShapefileSidecarFiles.apply(this, arguments);
}

export function replaceExtension(url, newExtension) {
  var baseName = basename(url);
  var extension = extname(url);
  var isUpperCase = extension === extension.toUpperCase();

  if (isUpperCase) {
    newExtension = newExtension.toUpperCase();
  }

  return "".concat(baseName, ".").concat(newExtension);
}

function basename(url) {
  var extIndex = url && url.lastIndexOf('.');
  return extIndex >= 0 ? url.substr(0, extIndex) : '';
}

function extname(url) {
  var extIndex = url && url.lastIndexOf('.');
  return extIndex >= 0 ? url.substr(extIndex + 1) : '';
}
