// @loaders.gl/shapefile/dist/esm/lib/parsers/parse-shp.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _awaitAsyncGenerator from "@babel/runtime/helpers/esm/awaitAsyncGenerator";
import _wrapAsyncGenerator from "@babel/runtime/helpers/esm/wrapAsyncGenerator";
import _asyncIterator from "@babel/runtime/helpers/esm/asyncIterator";
import BinaryChunkReader from '../streaming/binary-chunk-reader';
import { parseSHPHeader } from './parse-shp-header';
import { parseRecord } from './parse-shp-geometry';
var LITTLE_ENDIAN = true;
var BIG_ENDIAN = false;
var SHP_HEADER_SIZE = 100;
var SHP_RECORD_HEADER_SIZE = 12;
var STATE = {
  EXPECTING_HEADER: 0,
  EXPECTING_RECORD: 1,
  END: 2,
  ERROR: 3
};

var SHPParser = function () {
  function SHPParser(options) {
    _classCallCheck(this, SHPParser);

    this.options = options;
    this.binaryReader = new BinaryChunkReader({
      maxRewindBytes: SHP_RECORD_HEADER_SIZE
    });
    this.state = STATE.EXPECTING_HEADER;
    this.result = {
      geometries: []
    };
  }

  _createClass(SHPParser, [{
    key: "write",
    value: function write(arrayBuffer) {
      this.binaryReader.write(arrayBuffer);
      this.state = parseState(this.state, this.result, this.binaryReader, this.options);
    }
  }, {
    key: "end",
    value: function end() {
      this.binaryReader.end();
      this.state = parseState(this.state, this.result, this.binaryReader, this.options);

      if (this.state !== STATE.END) {
        this.state = STATE.ERROR;
        this.result.error = 'SHP incomplete file';
      }
    }
  }]);

  return SHPParser;
}();

export function parseSHP(arrayBuffer, options) {
  var shpParser = new SHPParser(options);
  shpParser.write(arrayBuffer);
  shpParser.end();
  return shpParser.result;
}
export function parseSHPInBatches(_x, _x2) {
  return _parseSHPInBatches.apply(this, arguments);
}

function _parseSHPInBatches() {
  _parseSHPInBatches = _wrapAsyncGenerator(_regeneratorRuntime.mark(function _callee(asyncIterator, options) {
    var parser, headerReturned, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _value, arrayBuffer;

    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            parser = new SHPParser(options);
            headerReturned = false;
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _context.prev = 4;
            _iterator = _asyncIterator(asyncIterator);

          case 6:
            _context.next = 8;
            return _awaitAsyncGenerator(_iterator.next());

          case 8:
            _step = _context.sent;
            _iteratorNormalCompletion = _step.done;
            _context.next = 12;
            return _awaitAsyncGenerator(_step.value);

          case 12:
            _value = _context.sent;

            if (_iteratorNormalCompletion) {
              _context.next = 27;
              break;
            }

            arrayBuffer = _value;
            parser.write(arrayBuffer);

            if (!(!headerReturned && parser.result.header)) {
              _context.next = 20;
              break;
            }

            headerReturned = true;
            _context.next = 20;
            return parser.result.header;

          case 20:
            if (!(parser.result.geometries.length > 0)) {
              _context.next = 24;
              break;
            }

            _context.next = 23;
            return parser.result.geometries;

          case 23:
            parser.result.geometries = [];

          case 24:
            _iteratorNormalCompletion = true;
            _context.next = 6;
            break;

          case 27:
            _context.next = 33;
            break;

          case 29:
            _context.prev = 29;
            _context.t0 = _context["catch"](4);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 33:
            _context.prev = 33;
            _context.prev = 34;

            if (!(!_iteratorNormalCompletion && _iterator["return"] != null)) {
              _context.next = 38;
              break;
            }

            _context.next = 38;
            return _awaitAsyncGenerator(_iterator["return"]());

          case 38:
            _context.prev = 38;

            if (!_didIteratorError) {
              _context.next = 41;
              break;
            }

            throw _iteratorError;

          case 41:
            return _context.finish(38);

          case 42:
            return _context.finish(33);

          case 43:
            parser.end();

            if (!(parser.result.geometries.length > 0)) {
              _context.next = 47;
              break;
            }

            _context.next = 47;
            return parser.result.geometries;

          case 47:
            return _context.abrupt("return");

          case 48:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[4, 29, 33, 43], [34,, 38, 42]]);
  }));
  return _parseSHPInBatches.apply(this, arguments);
}

function parseState(state) {
  var result = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var binaryReader = arguments.length > 2 ? arguments[2] : undefined;
  var options = arguments.length > 3 ? arguments[3] : undefined;

  while (true) {
    try {
      switch (state) {
        case STATE.ERROR:
        case STATE.END:
          return state;

        case STATE.EXPECTING_HEADER:
          var dataView = binaryReader.getDataView(SHP_HEADER_SIZE);

          if (!dataView) {
            return state;
          }

          result.header = parseSHPHeader(dataView);
          result.progress = {
            bytesUsed: 0,
            bytesTotal: result.header.length,
            rows: 0
          };
          result.currentIndex = 1;
          state = STATE.EXPECTING_RECORD;
          break;

        case STATE.EXPECTING_RECORD:
          while (binaryReader.hasAvailableBytes(SHP_RECORD_HEADER_SIZE)) {
            var recordHeaderView = binaryReader.getDataView(SHP_RECORD_HEADER_SIZE);
            var recordHeader = {
              recordNumber: recordHeaderView.getInt32(0, BIG_ENDIAN),
              byteLength: recordHeaderView.getInt32(4, BIG_ENDIAN) * 2,
              type: recordHeaderView.getInt32(8, LITTLE_ENDIAN)
            };

            if (!binaryReader.hasAvailableBytes(recordHeader.byteLength - 4)) {
              binaryReader.rewind(SHP_RECORD_HEADER_SIZE);
              return state;
            }

            var invalidRecord = recordHeader.byteLength < 4 || recordHeader.type !== result.header.type || recordHeader.recordNumber !== result.currentIndex;

            if (invalidRecord) {
              binaryReader.rewind(SHP_RECORD_HEADER_SIZE - 4);
            } else {
              binaryReader.rewind(4);
              var recordView = binaryReader.getDataView(recordHeader.byteLength);
              var geometry = parseRecord(recordView, options);
              result.geometries.push(geometry);
              result.currentIndex++;
              result.progress.rows = result.currentIndex - 1;
            }
          }

          if (binaryReader.ended) {
            state = STATE.END;
          }

          return state;

        default:
          state = STATE.ERROR;
          result.error = "illegal parser state ".concat(state);
          return state;
      }
    } catch (error) {
      state = STATE.ERROR;
      result.error = "SHP parsing failed: ".concat(error.message);
      return state;
    }
  }
}
