// @loaders.gl/shapefile/dist/es6/lib/parsers/deprecated/parse-dbf-atomic.js
import BinaryReader from '../../streaming/binary-reader';
const LITTLE_ENDIAN = true;
const DBF_HEADER_SIZE = 32;
export default function parseDbf(arrayBuffer, options) {
  const binaryReader = new BinaryReader(arrayBuffer);
  const loaderOptions = options.dbf || {};
  const {
    encoding
  } = loaderOptions;
  const fileHeaderView = binaryReader.getDataView(DBF_HEADER_SIZE);
  const header = parseDBFHeader(fileHeaderView);
  const {
    headerLength,
    recordLength,
    nRecords
  } = header;
  const textDecoder = new TextDecoder(encoding);
  const colHeaderView = binaryReader.getDataView(headerLength - DBF_HEADER_SIZE);
  const fields = parseColumnHeaders(colHeaderView, textDecoder);
  binaryReader.skip(1);
  return parseRows(binaryReader, fields, nRecords, recordLength, textDecoder);
}

function parseDBFHeader(headerView) {
  return {
    year: headerView.getUint8(1) + 1900,
    month: headerView.getUint8(2),
    day: headerView.getUint8(3),
    nRecords: headerView.getUint32(4, LITTLE_ENDIAN),
    headerLength: headerView.getUint16(8, LITTLE_ENDIAN),
    recordLength: headerView.getUint16(10, LITTLE_ENDIAN),
    languageDriver: headerView.getUint8(29)
  };
}

function parseColumnHeaders(view, textDecoder) {
  const nFields = (view.byteLength - 1) / 32;
  const fields = [];
  let offset = 0;

  for (let i = 0; i < nFields; i++) {
    const name = textDecoder.decode(new Uint8Array(view.buffer, view.byteOffset + offset, 11)).replace(/\u0000/g, '');
    fields.push({
      name,
      dataType: String.fromCharCode(view.getUint8(offset + 11)),
      fieldLength: view.getUint8(offset + 16),
      decimal: view.getUint8(offset + 17)
    });
    offset += 32;
  }

  return fields;
}

function parseRows(binaryReader, fields, nRecords, recordLength, textDecoder) {
  const rows = [];

  for (let i = 0; i < nRecords; i++) {
    const recordView = binaryReader.getDataView(recordLength - 1);
    binaryReader.skip(1);
    rows.push(parseRow(recordView, fields, textDecoder));
  }

  return rows;
}

function parseRow(view, fields, textDecoder) {
  const out = {};
  let offset = 0;

  for (const field of fields) {
    const text = textDecoder.decode(new Uint8Array(view.buffer, view.byteOffset + offset, field.fieldLength));
    out[field.name] = parseField(text, field.dataType);
    offset += field.fieldLength;
  }

  return out;
}

function parseField(text, dataType) {
  switch (dataType) {
    case 'B':
      return parseNumber(text);

    case 'C':
      return parseCharacter(text);

    case 'F':
      return parseNumber(text);

    case 'N':
      return parseNumber(text);

    case 'O':
      return parseNumber(text);

    case 'D':
      return parseDate(text);

    case 'L':
      return parseBoolean(text);

    default:
      throw new Error('Unsupported data type');
  }
}

function parseDate(str) {
  return Date.UTC(str.slice(0, 4), parseInt(str.slice(4, 6), 10) - 1, str.slice(6, 8));
}

function parseBoolean(value) {
  return /^[nf]$/i.test(value) ? false : /^[yt]$/i.test(value) ? true : null;
}

function parseNumber(text) {
  const number = parseFloat(text);
  return isNaN(number) ? null : number;
}

function parseCharacter(text) {
  return text.trim() || null;
}
